homebridge/thermostat.ts

```diff
@@ -0,0 +1,189 @@
+          (mode === 'heat' || mode === 'aux')
+        ) {
+          // The current temperature is lower than the target temperature,
+          // and the thermostat is set to 'heat' or 'aux' (emergency heat), so the thermostat is heating
+          return Characteristic.CurrentHeatingCoolingState.HEAT
+        }
+        // The current temperature is either higher or lower than the target temperature,
+        // but the current thermostat mode would only to increase the difference,
+        // so the thermostat is neither heating nor cooling
+        return Characteristic.CurrentHeatingCoolingState.OFF
+      },
+    })
+
+    this.registerCharacteristic({
+      characteristicType: Characteristic.TargetHeatingCoolingState,
+      serviceType: Service.Thermostat,
+      getValue: ({ mode }) => {
+        switch (mode) {
+          case 'off':
+            return Characteristic.TargetHeatingCoolingState.OFF
+          case 'heat':
+          case 'aux':
+            return Characteristic.TargetHeatingCoolingState.HEAT
+          case 'cool':
+            return Characteristic.TargetHeatingCoolingState.COOL
+        }
+      },
+      setValue: (targetHeatingCoolingState: number) => {
+        const mode: ThermostatMode | undefined = (():
+          | ThermostatMode
+          | undefined => {
+          switch (targetHeatingCoolingState) {
+            case Characteristic.TargetHeatingCoolingState.OFF:
+              return 'off'
+            case Characteristic.TargetHeatingCoolingState.HEAT:
+              return 'heat'
+            case Characteristic.TargetHeatingCoolingState.COOL:
+              return 'cool'
+            default:
+              return
+          }
+        })()
+        if (!mode) {
+          this.logger.error(
+            `Couldn’t match ${targetHeatingCoolingState} to a recognized mode string.`
+          )
+          return
+        }
+        this.logger.info(`Setting ${this.device.name} mode to ${mode}`)
+
+        return this.device.setInfo({ device: { v1: { mode } } })
+      },
+    })
+    // Only allow 'TargetHeatingCoolingState's which can be mapped to Ring modes
+    // Specifically, this omits .AUTO
+    this.getService(Service.Thermostat)
+      .getCharacteristic(Characteristic.TargetHeatingCoolingState)
+      .setProps({
+        validValues: [
+          Characteristic.TargetHeatingCoolingState.OFF,
+          Characteristic.TargetHeatingCoolingState.HEAT,
+          Characteristic.TargetHeatingCoolingState.COOL,
+        ],
+      })
+
+    this.registerCharacteristic({
+      characteristicType: Characteristic.CurrentTemperature,
+      serviceType: Service.Thermostat,
+      getValue: async () => {
+        const temperature = await this.getTemperatureFromComponentDevices()
+        if (!temperature) {
+          return
+        }
+        // Documentation: https://developers.homebridge.io/#/characteristic/CurrentTemperature
+        // 'Characteristic.CurrentTemperature' supports 0.1 increments
```

Comment:
https://github.com/chrisjshull/homebridge-nest/blob/master/lib/nest-thermostat-accessory.js#L70 - looks like you _can_ adjust this step size, though I'm not sure what step size ring supports so `0.1` is probably just fine.

Reply:
Oh interesting, thanks for the tip! I wonder whether the issue alluded to in [the comment](https://github.com/chrisjshull/homebridge-nest/blob/a2165e5b572dfa488408ef09192099852d2619d9/lib/nest-thermostat-accessory.js#L59-L62) a few lines above will affect us:
```JS
/*
 * Only allow 0.5 increments for Celsius temperatures. HomeKit is already limited to 1-degree increments in Fahrenheit,
 * and setting this value for Fahrenheit will cause HomeKit to incorrectly round values when converting from °F to °C and back.
 */
```

A `0.1` degrees celsius step is about `0.18` degrees Fahrenheit at typical room temperatures. The Home app only supports round-number set points. I wouldn’t be surprised if it supported half-degree set points in the future, but I doubt it ever supports anything finer than that (e.g. `0.18`).

> I'm not sure what step size ring supports so 0.1 is probably just fine.

I also don’t know. I propose we leave the step as-is for now, and revisit in the future if someone reports a disparity between the Fahrenheit number displayed on their wall unit and the Fahrenheit number displayed in the Home app (the latter converted from the celsius value included in the API response).